### DISCUSSION TITLE
Replacing default icons with custom svg icons?

### DISCUSSION THREAD

**Discussion by thomas-kuhlmann** [quasarframework/quasar#11713] [General - Components / Directives / etc] [answered]
Hi,

I am fairly new to Quasar and even though the documentation is superb, I seem to be a bit stuck - is there a quasar way to replace existing system symbols (I'm using vite)? I understand that I could map my icon library (Remixicons) via an icon set file, but it looks as if that mapping is based on the class names? So I would have to install the webfont, not just use the SVG icons?

If anyone has a good tutorial on how to replace default system icons with equivalent icons from a custom icon library like remixicons, it'd be greatly appreciated. Or if anybody actually has already done the work of mapping remixicons to quasar and would be willing to share this, I'd be in your debt! 
I have already started using https://github.com/hawkeye64/quasar-extras-svg-icons to use custom SVG icons in q-icon, but how to replace i.e. chevrons in a date picker with SVG versions seem to be beyond me...

Thanks a lot in advance for any pointers!

**Comment by hawkeye64** [comment]
If I get some time this week, I'll make a project and put my findings in the docs. How quickly do you need this? Is it a blocker right now?

**Comment by hawkeye64** [comment, marked as answer]
@ThomasKuhlmann I have to say, this is fairly easy.

here is my `script setup` that changes the chevron used on QExpansionItem (with a remix icon). I used an icon that would be recognizable as NOT the default one used.

```js
<script setup>
  import { useQuasar } from 'quasar'
  import { remArrowDownCircleFill } from 'quasar-extras-svg-icons/remix-icons'

  const $q = useQuasar()
  $q.iconSet.expansionItem.icon = remArrowDownCircleFill
</script>
```

And, a little video showing it in use:

https://user-images.githubusercontent.com/10262924/146585677-dd792f6c-8d5a-422b-a0be-990a4a91ccd0.mp4

**Reply by thomas-kuhlmann** [reply]
Thanks a lot for the detailed walkthrough - this definitely clarified how the system works (and you are right, this was easy)! Seeing that you can simply replace the class name with an SVG file means that I could just import and do the mapping in a central file and import it in the [quasar config as per your guideline ](https://quasar.dev/options/quasar-icon-sets#quasar-vite-plugin-way)

Here's the mapping I have done for replacing the material design icons with Remixicons (in case somebody wants to save some time)

```
import {
  remArrowUpSLine,
  remArrowDownSLine,
  remArrowLeftSLine,
  remArrowRightSLine,
  remArrowLeftSFill,
  remArrowRightSFill,
  remArrowUpLine,
  remArrowDownLine,
  remArrowLeftLine,
  remArrowRightLine,
  remArrowDropDownLine,
  remInformationLine,
  remAlertLine,
  remErrorWarningLine,
  remCheckboxCircleLine,
  remRestartLine,
  remCheckboxBlankCircleFill,
  remCloseCircleLine,
  remCheckLine,
  remTimeLine,
  remCalendarEventLine,
  remStrikethrough,
  remItalic,
  remBold,
  remUnderline,
  remListUnordered,
  remListOrdered,
  remSubscript,
  remSuperscript,
  remLink,
  remFullscreenLine,
  remDoubleQuotesL,
  remAlignLeft,
  remAlignCenter,
  remAlignRight,
  remAlignJustify,
  remPrinterLine,
  remIndentDecrease,
  remIndentIncrease,
  remFormatClear,
  remFontColor,
  remFontSize,
  remSeparator,
  remArrowGoBackLine,
  remArrowGoForwardLine,
  remHeading,
  remH1,
  remH2,
  remH3,
  remH4,
  remH5,
  remH6,
  remCodeLine,
  remNumber1,
  remNumber2,
  remNumber3,
  remNumber4,
  remNumber5,
  remFontSize2,
  remNumber6,
  remNumber7,
  remAddLine,
  remCloseLine,
  remStarLine,
  remPencilLine,
  remArrowDropRightLine,
  remUploadCloud2Line,
  remDeleteRow,
  remCheckDoubleLine,
} from "quasar-extras-svg-icons/remix-icons";

export default {
  name: "remixicons",
  type: {
    positive: remCheckboxCircleLine,
    negative: remAlertLine,
    info: remInformationLine,
    warning: remErrorWarningLine,
  },
  arrow: {
    up: remArrowUpLine,
    right: remArrowRightLine,
    down: remArrowDownLine,
    left: remArrowLeftLine,
    dropdown: remArrowDropDownLine,
  },
  chevron: {
    left: remArrowLeftSLine,
    right: remArrowRightSLine,
  },
  // colorPicker: {
  //   spectrum: "mdi-gradient-vertical",
  //   tune: "mdi-tune",
  //   palette: "mdi-palette-swatch",
  // },
  pullToRefresh: {
    icon: remRestartLine,
  },
  carousel: {
    left: remArrowLeftSLine,
    right: remArrowRightSLine,
    up: remArrowUpSLine,
    down: remArrowDownSLine,
    navigationIcon: remCheckboxBlankCircleFill,
  },
  chip: {
    remove: remCloseCircleLine,
    selected: remCheckLine,
  },
  datetime: {
    arrowLeft: remArrowLeftSLine,
    arrowRight: remArrowRightSLine,
    now: remTimeLine,
    today: remCalendarEventLine,
  },
  editor: {
    bold: remBold,
    italic: remItalic,
    strikethrough: remStrikethrough,
    underline: remUnderline,
    unorderedList: remListUnordered,
    orderedList: remListOrdered,
    subscript: remSubscript,
    superscript: remSuperscript,
    hyperlink: remLink,
    toggleFullscreen: remFullscreenLine,
    quote: remDoubleQuotesL,
    left: remAlignLeft,
    center: remAlignCenter,
    right: remAlignRight,
    justify: remAlignJustify,
    print: remPrinterLine,
    outdent: remIndentDecrease,
    indent: remIndentIncrease,
    removeFormat: remFormatClear,
    formatting: remFontColor,
    fontSize: remFontSize,
    align: remAlignLeft,
    hr: remSeparator,
    undo: remArrowGoBackLine,
    redo: remArrowGoForwardLine,
    heading: remHeading,
    heading1: remH1,
    heading2: remH2,
    heading3: remH3,
    heading4: remH4,
    heading5: remH5,
    heading6: remH6,
    code: remCodeLine,
    size: remFontSize2,
    size1: remNumber1,
    size2: remNumber2,
    size3: remNumber3,
    size4: remNumber4,
    size5: remNumber5,
    size6: remNumber6,
    size7: remNumber7,
    font: remFontSize,
    viewSource: remCodeLine,
  },
  expansionItem: {
    icon: remArrowDownSLine,
    denseIcon: remArrowDropDownLine,
  },
  fab: {
    icon: remAddLine,
    activeIcon: remCloseLine,
  },
  field: {
    clear: remCloseCircleLine,
    error: remErrorWarningLine,
  },
  pagination: {
    first: remArrowLeftSFill,
    prev: remArrowLeftSLine,
    next: remArrowRightSLine,
    last: remArrowRightSFill,
  },
  rating: {
    icon: remStarLine,
  },
  stepper: {
    done: remCheckLine,
    active: remPencilLine,
    error: remAlertLine,
  },
  tabs: {
    left: remArrowLeftSLine,
    right: remArrowRightSLine,
    up: remArrowUpSLine,
    down: remArrowDownSLine,
  },
  table: {
    arrowUp: remArrowUpLine,
    warning: remAlertLine,
    firstPage: remArrowLeftSFill,
    prevPage: remArrowLeftSLine,
    nextPage: remArrowRightSLine,
    lastPage: remArrowRightSFill,
  },
  tree: {
    icon: remArrowDropRightLine,
  },
  uploader: {
    done: remCheckLine,
    clear: remCloseLine,
    add: remAddLine,
    upload: remUploadCloud2Line,
    removeQueue: remDeleteRow,
    removeUploaded: remCheckDoubleLine,
  },
};
```

**Reply by hawkeye64** [reply]
Beatiful!

**Reply by alensiljak** [reply]
@ThomasKuhlmann, how did you add the custom icon set definition to Quasar? 
`framework.iconSet` parameter in quasar.config.js expects a string and then looks for it at a location where all the other pre-defined icon sets sit. 
I was previously checking the [mdi svg](https://github.com/quasarframework/quasar/blob/dev/ui/icon-set/svg-material-icons.js) definition and that's the syntax that works for including it in the quasar.config.js. But still I see no clear way to state that this is the icon set I want as default.

**Comment by ssuess** [comment]
I feel like I am missing something here, I tried to do the above, but to be clear: I don't want to dynamically replace quasar icons with mine, I just want to define mine as the default set from the get go, is this possible? I have tried several combinations of boot file like the following:
```
import { boot } from 'quasar/wrappers'
import { Quasar } from 'quasar'
import iconSet from '../components/iconsetmap'

export default boot(({ app }) => {
  app.use(Quasar, {
    iconSet
  })
})
```
where iconsetmap is a file like the above with the imported icon names next to the quasar element names, but it does not work. None of the svg icons I have defined are showing up.

I would like to do this WITHOUT woff or other font files, only svg...is this even possible?

**Reply by alensiljak** [reply]
I assume this is exactly what Thomas did. You need to define an icon set (see, for example: https://github.com/quasarframework/quasar/blob/dev/ui/icon-set/svg-material-icons.js or the one just two posts above from yours). Then include it in the Quasar config. I'm about to try.

**Reply by ssuess** [reply]
@alensiljak did you ever get this to work? I can't. My imports give me no errors, the mapping I have triplechecked and it looks good, but all I have is blank spaces where the Quasar component icons should be.  :(

**Reply by alensiljak** [reply]
@ssuess, unfortunately, no. I gave up and simply used a pre-defined set. It does not seem to work or the docs are not covering it all.

**Comment by ssuess** [comment]
This is driving me crazy. Can anyone walk us through the steps necessary in every file? Here is what I have so far:
1. Boot file:
```
import { boot } from 'quasar/wrappers'
import { Quasar } from 'quasar'
import iconSet from '../components/iconsetmap.js'

export default boot(({ app }) => {
  app.use(Quasar, {
    iconSet
  })
})
```
2. `quasar.config.js` has boot file enabled in boot section, otherwise nothing else. (nothing in framework and nothing in extras at the moment)
3. my `iconsetmap.js` file looks the same as the one posted above by @ThomasKuhlmann (except for fontawesome-pro replacements)
4. my imported (by iconsetmap.js above) icon def file is using the proper format for each icon (ie `export const fasSmile = 'M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm80 168c17.7 0 32 14.3 32 32s-14.3 32-32 32-32-14.3-32-32 14.3-32 32-32zm-160 0c17.7 0 32 14.3 32 32s-14.3 32-32 32-32-14.3-32-32 14.3-32 32-32zm194.8 170.2C334.3 380.4 292.5 400 248 400s-86.3-19.6-114.8-53.8c-13.6-16.3 11-36.7 24.6-20.5 22.4 26.9 55.2 42.2 90.2 42.2s67.8-15.4 90.2-42.2c13.4-16.2 38.1 4.2 24.6 20.5z|0 0 496 512'`)

What on earth am I still missing that is resulting is blank icons in all the quasar components???

**Comment by ssuess** [comment]
Since no one seems to know how to do this (or an answer isn't forthcoming), I guess I am going to go with the inelegant solution of creating a mixin/function and import it everywhere. This works to reduce my built code size, but is a lot of rewriting. Since I am using Options API in this project, I will create a mixin that looks like this:
```
import { useQuasar } from 'quasar'
import * as rFonts from '../components/light.js' // contains icon defs as above ie export const fasSmile = 'M248...
export default {
  data () {
    return {
    }
  },
  beforeMount () {
    const $q = useQuasar()
    // quasar icon replacements here
    $q.iconSet.expansionItem.icon = rFonts.falChevronDown
    $q.iconSet.type.positive = rFonts.falCheck
    $q.iconSet.type.negative = rFonts.falExclamationTriangle
    ...
    // other icons defined here
    this.falSmile = rFonts.falSmile
    this.falAngleRight = rFonts.falAngleRight
    this.falAngleDown = rFonts.falAngleDown
    ...
    }
}
```
and then just import this mixin wherever needed in other vue files. Not at all the solution I was hoping for, but it works. hope this helps someone else struggling with this.

**Reply by yusufkandemir** [reply]
https://github.com/quasarframework/quasar/discussions/11713#discussioncomment-4474671

**Comment by yusufkandemir** [comment]
https://quasar.dev/options/quasar-icon-sets#changing-icon-set-dynamically
The way to change icon sets at runtime is using `$q.iconSet.set(iconSetDefinition)`
```js
import { ref } from 'vue'
import { useQuasar } from 'quasar'
import faIconSet from 'quasar/icon-set/fontawesome-v6'
import myIconSet from 'src/assets/my-icon-set'

const $q = useQuasar()
const isMySet = ref(false)
function toggleIconSet() {
  isMySet.value = !isMySet.value
  $q.iconSet.set(isMySet.value ? myIconSet : faIconSet)
}
```

If you want to set a custom icon set once, you can do it in `quasar.config.js`:
```js
framework: {
  iconSet: 'src/assets/my-icon-set' // path to your custom icon set file
}
```

A custom icon set file would look like this: https://github.com/quasarframework/quasar/blob/dev/ui/icon-set/svg-mdi-v7.mjs

**Reply by ssuess** [reply]
Thanks, this helps a bit, but  trying to use any of these icons anywhere but in the defaults fails. For example, a non modified table will correctly show forward and back paging arrows, but if I try to replace them with a template (as I have before with no issues when using webfonts instead of svgs) with ones that are defined in the same file, it fails with errors like `Property "falChevronRight" was accessed during render but is not defined on instance.`
Even if I add it to the export iconSet lines like: `falChevronRight: rFonts.falChevronRight` Same thing if I try to use these in `q-btn` or other components.

**Reply by yusufkandemir** [reply]
@ssuess I couldn't understand anything. Please explain clearly, with an example code, show where exactly it fails, etc. There shouldn't be any property called `falChevronRight` in the icon sets, it would be like `carousel: { right: falChevronRight, ... }`, not the other way. So, not sure what you are doing.

**Reply by ssuess** [reply]
@yusufkandemir sure, sorry I wasn't more clear. If I just use the defaults which are set (exactly like `carousel: { right: falChevronRight, ...` and don't customize any of the components or icons, this works. But if I try to use (for example) the templates for tables like so it does not work:
```
<template v-slot:pagination="scope">
        <q-btn
          v-if="scope.pagesNumber > 2"
          :icon="falChevronDoubleLeft"
          size="sm"
          color="grey-8"
          round
          dense
          flat
          :disable="scope.isFirstPage"
          @click="scope.firstPage"
        />
        ...
      </template>
```      
Whereas previously this WOULD work:
```
<template v-slot:pagination="scope">
        <q-btn
          v-if="scope.pagesNumber > 2"
          icon="fal fa-chevron-double-left"
          size="sm"
          color="grey-8"
          round
          dense
          flat
          :disable="scope.isFirstPage"
          @click="scope.firstPage"
        />
        ...
      </template>
```
So I tried adding values to my iconset file like:
```
export default {
  name: falIcons,
  falChevronDoubleLeft: rFonts.falChevronDoubleLeft,
  type: {
    positive: rFonts.falCheck,
    negative: rFonts.falExclamationTriangle,
    info: rFonts.falInfoCircle,
    warning: rFonts.falExclamation
  },
...
```
but that did not work. I don't want to have to create and import a mixin for the multitude of files that I have that contain these types of customizations, but I can't see any other way. Getting rid of the webfont in favor of svgs is not at all drop in.

**Reply by ssuess** [reply]
I ended up doing just that (using a mixin for anything not in the base iconSet) with a bunch of search and replaces, and now everything is working the way I want.

**Reply by yusufkandemir** [reply]
Putting new stuff into the icon set doesn't do magical stuff, not sure what made you think that it would work. There is nothing similar to that in the docs. If you do `:icon="falChevronDoubleLeft"`, you got to import it and make it available to the template. The docs covers this: https://quasar.dev/vue-components/icon#svg-usage
